src/core/stack_switching/suspenders.mjs

```diff
@@ -0,0 +1,178 @@
+  const wrapped_type = structuredClone(orig_type);
+  wrapped_type.parameters.unshift("externref");
+  const orig_sig = ts.addWasm(orig_type);
+  const wrapped_sig = ts.addWasm(wrapped_type);
+  mod.addSection(ts);
+
+  const imports = new ImportSection();
+  imports.addGlobal("s", "externref");
+  const orig = imports.addFunction("i", orig_sig);
+  mod.addImportSection(imports);
+  mod.setExportType(wrapped_sig);
+
+  const code = new CodeSection();
+  code.local_get(0);
+  code.global_set(0);
+  for (let i = 1; i < wrapped_type.parameters.length; i++) {
+    code.local_get(i);
+  }
+  code.call(orig);
+  mod.addSection(code);
+  return mod.generate();
+}
+
+const promisingModuleMap = new Map();
+function getPromisingModule(orig_type) {
+  const type_str = wasmTypeToString(orig_type);
+  if (promisingModuleMap.has(type_str)) {
+    return promisingModuleMap.get(type_str);
+  }
+  const module = createPromisingModule(orig_type);
+  promisingModuleMap.set(type_str, module);
+  return module;
+}
+
+const promisingFunctionMap = new WeakMap();
+/**
+ * This creates a wrapper around wasm_func that receives an extra suspender
+ * argument and returns a promise. The suspender is stored into suspenderGlobal
+ * so it can be used by syncify (see wrap_syncifying.wat)
+ *
+ * For unit testing, allow passing suspenderGlobal as an argument.
+ */
+export function createPromising(wasm_func) {
+  if (promisingFunctionMap.has(wasm_func)) {
+    return promisingFunctionMap.get(wasm_func);
+  }
+  const type = WebAssembly.Function.type(wasm_func);
+  const module = getPromisingModule(type);
+  const instance = new WebAssembly.Instance(module, {
+    e: { i: wasm_func, s: suspenderGlobal },
+  });
+  const result = new WebAssembly.Function(
+    { parameters: type.parameters, results: ["externref"] },
+    instance.exports.o,
+    { promising: "first" },
+  );
+  promisingFunctionMap.set(wasm_func, result);
+  return result;
+}
+
+export let suspenderGlobal;
+try {
+  suspenderGlobal = new WebAssembly.Global(
+    { value: "externref", mutable: true },
+    null,
+  );
+} catch (e) {}
+
+let validSuspender;
+
+/**
+ * This sets up syncify to work.
+ *
+ * We need to make:
+ *
+ * - suspenderGlobal where we store the suspender object
+ *
+ * - promisingApplyHandler which calls a Python function with stack switching
+ *   enabled (used in callPyObjectKwargsSuspending in pyproxy.ts)
+ *
+ * - the syncifyHandler which uses suspenderGlobal to suspend execution, then
+ *   awaits a promise, then resumes execution and returns the promise result
+ *   (used by hiwire_syncify)
+ *
+ * If the creation of these fails because JSPI is missing, then we set it up so
+ * that callKwargsSyncifying and hiwire_syncify will always raise errors and
+ * everything else can work as normal.
+ */
+export function initSuspenders() {
+  // It would be nice to have a better way to feature detect wasm stack
+  // switching than this. For now I haven't come up with anything better.
```

Comment:
[This](https://github.com/GoogleChromeLabs/wasm-feature-detect/blob/main/src/detectors/jspi/index.js) doesn't work ?

Reply:
BTW that whole library looks quite nice (and lightweight)  https://github.com/GoogleChromeLabs/wasm-feature-detect maybe we should be using it?

Reply:
It does work. Though I'm a bit unclear on what the point of `WebAssembly.Suspender` is. I opened an issue about it on the proposal:
https://github.com/WebAssembly/js-promise-integration/issues/21